derive_stringset/src/lib.rs

```diff
@@ -0,0 +1,69 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, DeriveInput};
+
+#[proc_macro_derive(StringSet)]
+pub fn derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    let name = input.ident;
+    let fields = if let syn::Data::Struct(syn::DataStruct {
+        fields: syn::Fields::Named(syn::FieldsNamed { ref named, .. }),
+        ..
+    }) = input.data
+    {
+        named
+    } else {
+        // A struct with named fields is the only supported input
+        unimplemented!();
+    };
+
+    let recurse_match_arms = fields.iter().map(|f| {
+        let name = &f.ident;
+        quote! {
+            stringify!(#name) => {
+                self.#name.string_set(topic_parts, value)?;
+                Ok(())
+            }
+        }
+    });
+
+    let direct_set_match_arms = fields.iter().map(|f| {
+        let name = &f.ident;
+        quote! {
+            stringify!(#name) => {
+                self.#name = serde_json_core::from_str(value)
+                    .map_err(|_|{()})?.0;
+                Ok(())
+            }
+        }
+    });
+
+    let expanded = quote! {
+        impl StringSet for #name {
```

Comment:
A weirdness I noticed is that it seems like the _user_ crate needs to add `serde-json-core` to their dependencies or things fail, since we're injecting calls to it into their code. Idk the best way around this yet.